### ISSUE TITLE
Proposal for a blocklist to mitigate the 255 device limit

### ISSUE THREAD

**Issue by hughb8on** [stas-demydiuk/domoticz-zigbee2mqtt-plugin#594]
TLDR: this is a proposal for the plugin to support an optional blocklist file to suppress handling of some devices and/or sub-devices.

Like others with lots of zigbee kit I've been running up against the Domoticz framework's limit of 255 devices per plugin.

The option added last year to suppress signal and/or cell and btperc devices helped considerably. However the Universal Adaptor approach, which is basically "if zigbee2mqtt.io exposes it, let's create a device for it" is making things worse again.

Though it's not the fault of the plugin, many of the newly-exposed attributes of devices are completely redundant. For example, there is no point to "tamper" devices for Sonoff sensors which don't even have hardware to detect tampering (go figure!); the temp sensors which Xiaomi/Aqara motion and contact sensors expose appear to report nonsense values;  a single Temp/Humidity/Pressure device can result in nine (yes really!) Domoticz devices And when the novelty of "effec" devices wears off you might rather not have those either. I could go on and on ..

I've been looking for a way to make the plugin more discriminating in its handling of these sub-devices and I'd like to propose a solution.

In my personal, temporary fork of the plugin I've added, and have been testing, code to read in an optional file from the plugin folder, which I've called blocklist.txt.

Each line in the file is a device-id or a fragment of one; each will therefore most probably be one of the following: -
    a full device-id as seen in the ID column in Domoticz's Setup/Devices tab, e.g. 0x00124b001f841999_cell
    a device suffix, e.g. cell, which will suppress all cell devices
    an IEEE Address,e.g. 0x00124b001f841999 which will suppress that zigbee device completely

Note that you aren't restricted to just those id fragments, any will do. For example if for some reason you wanted to suppress all your Sonoff devices you could add a line with just 0x00124b0 on it. I considered using regex to allow really sophisticated matching but in the end I couldn't see anybody needing it, there would be an overhead and python-regex probably isn't even present on most people's systems.

I've added the code to read the file to domoticz.py. This is called during the init of devices_manager.py. A global list variable is created called "block_list". This is then examined in device.py:register and device.py:handle_message to decide whether a particular device should be suppressed (i.e. not created and not updated, respectively).

Devices which are suppressed by the blocklist but which already exist will no longer be updated. If they are deleted from Domoticz the plugin will not recreate them. The Custom Zigbee2MQTT tab is not affected so you will still be able to see your full zigbee mesh there.

The blocklist file does not have to exist; is is entirely optional and this is therefore not a breaking change. Indeed I would expect most users won't even need to know the facility exists..

Note that the ability to suppress signal, cell and btperc devices using the blocklist makes the "Track Link Quality" and "Use Battery Devices" parameters on the plugin's Hardware page redundant, but there is no negative consequence if devices are suppressed "twice".

I'd like to know what others think, particularly @stas-demydiuk of course. My opinion is that even if the Domoticz device limit were raised at some point in the future it would still be useful for users with large numbers of devices to be able to reduce the clutter.

This is clearly not the only way to achieve device suppression, it's just the way I decided to go. I have little knowledge of the overall design of the plugin so I may well have missed hidden implications of my approach. Also, I am brand new to Python coding and it probably shows!

I'd very much welcome comments on the idea and hope something along these lines can be added to the plugin.

**Comment by Anjerlaan** [comment]
In the frontend of Zigbee2mqtt under settings/experimental settings is also a blocklist/whitelist available.
Is that the same of what you mean?

**Comment by hughb8on** [comment]
The blocklist facility in zigbee2mqtt.io is very different. It used to be called “ban”, because that is what it does  - it prevents a zigbee device from joining the network at all. 

My suggestion is to allow the suppression of unwanted features exposed by devices.

**Comment by Anjerlaan** [comment]
Is the 255 maximum devices only a Domoticz issue or also for Home Assistant?
if this is the case, wouldn’t it be an idea to implement this enhancement at the Zigbee2mqtt side?

**Comment by hughb8on** [comment]
It’s a restriction of the Domoticz plugin framework, affecting all plugins. There is a an ongoing discussion about it in one of the open issues here - you might like to read it.
You’d need to do your own research into the limits of Home Assistant, I don’t use it.
I think it would be a bad idea to fix a Domoticz problem by changing zigbee2mqtt.io.

**Comment by jeroenkl** [comment]
> The blocklist facility in zigbee2mqtt.io is very different. It used to be called “ban”, because that is what it does - it prevents a zigbee device from joining the network at all.
> 
> My suggestion is to allow the suppression of unwanted features exposed by devices.

Totally agree with your suggestion.

**Comment by dobber81** [comment]
I would love to see this sollution implemented by stas-demydiuk. An easy fix to a very annoying problem. I could easily cut away around 70% of the zigbee devices in my domoticz setup. 
Right now I'm closing in on the 255 device limit. With this fix it would be around 80 devices that I actually have use for. At the same time making the domoticz database file less occupied by garbage.

**Comment by heggink** [comment]
This looks like a workable solution but I could live with something probably more simple:
1) use domoticz for what it's meant to do: control and read primary device values (on/off, temp, battery level so domoticz can report on that, ..)
2) use the z2m control page to report and control secondary stuff (link level, power_on_behavior, color_temp_startup maybe) and keep these out of domoticz as these rarely require changes and can be set from z2m directly
3) add a 'simple' switch to the plugin settings to create separate devices or combined devices for the likes of temp/hum/baro

That way, you don't need to maintain any list and the number of devices (and device updates) is minimal. 

Thoughts?

**Comment by hughb8on** [comment]
As I see it, implementing the suppression of unwanted device features boils down to two things: firstly an interface for the user to choose what they want to suppress, and secondly somewhere to store those choices.

I think an optional blocklist file in the plugin folder achieves both of those things in a very simple way. 

Most users will have no need to even know about the option to create such a file.

More sophisticated users will already be well-used to the idea of editing configuration files to achieve more complicated things than a tidy web interface can present. 

For example, users who never find themselves needing to edit zigbee2mqtt’s configuration.yaml are unlikely to need a blocklist file, but I think those who do will welcome the option to create one.

**Comment by stas-demydiuk** [comment]
Blacklist support has been added using plugin settings page. Now there is one more parameter to setup the blacklist.

![image](https://user-images.githubusercontent.com/2734836/111068413-512f0780-84d1-11eb-9bbf-8e6f3a8bc6dd.png)

Blacklist could contain several entries devided by semi-colon. Each blacklist entry should be valid [Python regexp](https://docs.python.org/3/library/re.html#regular-expression-syntax). For example:
- Single item of a device - use full Device ID (ieee addr + alias) to block it i.e. `0x842e14fffe13971a_state`
- All items of a device - use Regexp to block all logical devices (aliases) of a specified zigbee device (ieee address) i.e. `0x842e14fffe13971a_.*`
- Specific item of all devices - use Regexp to block all devices (ieee address) with a specific alias i.e. `.*_state`

Any feedback is highly appriciated

**Comment by jeroenkl** [comment]
> Blacklist support has been added using plugin settings page. Now there is one more parameter to setup the blacklist.
> 
> ![image](https://user-images.githubusercontent.com/2734836/111068413-512f0780-84d1-11eb-9bbf-8e6f3a8bc6dd.png)
> 
> Blacklist could contain several entries devided by semi-colon. Each blacklist entry should be valid [Python regexp](https://docs.python.org/3/library/re.html#regular-expression-syntax) (the most common regexp entity I guess would be `.*` which means any number of symbols). For example:
> 
> * Single item of a device - use full Device ID (ieee addr + alias) to block it i.e. `0x842e14fffe13971a_state`
> * All items of a device - use Regexp to block all logical devices (aliases) of a specified zigbee device (ieee address) i.e. `0x842e14fffe13971a_.*`
> * Specific item of all devices - use Regexp to block all devices (ieee address) with a specific alias i.e. `.*_state`
> 
> Any feedback is highly appriciated

Extra Option also to choose 'whitelist'  instead of 'blacklist'?
I'll test the solution above, thanks!!

**Comment by stas-demydiuk** [comment]
No available options left :) Domoticz supports up to 6 custom parameters per plugin

**Comment by dobber81** [comment]
Looks great! I just dropped 40+ devices that I have no need for.

**Comment by jeroenkl** [comment]
> No available options left :) Domoticz supports up to 6 custom parameters per plugin

no problem ;-)

**Comment by jeroenkl** [comment]
Hmm, it is not working for me. The black_listed devices are still updated....
Is this line ok?
![image](https://user-images.githubusercontent.com/37976568/111070825-c13e8d00-84d3-11eb-9dc3-7127d0591c81.png)

**Comment by dobber81** [comment]
Try removing the devices. They will not be readded.

**Comment by jeroenkl** [comment]
> Try removing the devices. They will not be readded.

Yes, you're right. Thanks

**Comment by hughb8on** [comment]
I am finding that if the blacklist parameter is left blank then *all* devices are blacklisted.

**Comment by hughb8on** [comment]
.. and a suggestion .. please consider allowing the word 'heartbeat' as  one of the values. 

That would please those who who don't like seeing heartbeat messages being recorded, very confusingly, as motion detections, and lights being shown as having been turned on when they haven't.

**Comment by martin-g-it** [comment]
nice, works like a charm!

**Comment by felix63** [comment]
> .. and a suggestion .. please consider allowing the word 'heartbeat' as one of the values.
> 
> That would please those who who don't like seeing heartbeat messages being recorded, very confusingly, as motion detections, and lights being shown as having been turned on when they haven't.

How do you mean? How should you add heartbeat?

**Comment by hughb8on** [comment]
By typing it in the box along with any blocklisted device names.
